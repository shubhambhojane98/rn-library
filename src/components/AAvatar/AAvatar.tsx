import React, {FC} from 'react';
import {StyleSheet} from 'react-native';
import {moderateScale} from 'react-native-size-matters';
import {defaultScale} from '../../utils/Common';
import AImage from '../AImage/AImage';
import ATypography from '../ATypography/ATypography';
import {TextAlignment, TypographyVariant} from '../ATypography/ATypographyEnum';
import {withTheme, useTheme} from '../../core/theming';
import type {Theme} from '../../utils/types';

interface Prop {
  source?: object;
  sourceUri?: string;
  size?: number;
  backgroundColor?: string;
  resizeMode?: any;
  marginTop?: number;
  marginRight?: number;
  marginBottom?: number;
  marginLeft?: number;
  cache?: any;
  label?: string;
  textVariant?: TypographyVariant;
  fontSize?: number;
  borderRadius?: number;
  theme: Theme;
}

const defaultSize = 64;
const defaultMargin = 0;

const AAvatar: FC<Prop> = ({
  source,
  sourceUri,
  size,
  backgroundColor,
  resizeMode,
  marginTop,
  marginRight,
  marginBottom,
  marginLeft,
  cache,
  label,
  textVariant,
  fontSize,
  borderRadius,
}) => {
  const radiusValue =
    borderRadius !== undefined ? borderRadius : defaultSize / 2;

  const {colors} = useTheme();
  const bgColor = backgroundColor ? backgroundColor : colors.primary;

  const stylesWithProp = styles({colors});

  return (
    <>
      {label && (
        <ATypography
          children={label}
          textAlign={TextAlignment.Center}
          variant={textVariant}
          fontSize={fontSize}
          style={{
            ...stylesWithProp.labelStyle,
            lineHeight: size || defaultSize,
            width: size || defaultSize,
            height: size || defaultSize,
            borderRadius: radiusValue,
            backgroundColor: bgColor || colors.lightgrey,
            marginTop: moderateScale(marginTop || defaultMargin, defaultScale),
            marginRight: moderateScale(
              marginRight || defaultMargin,
              defaultScale,
            ),
            marginBottom: moderateScale(
              marginBottom || defaultMargin,
              defaultScale,
            ),
            marginLeft: moderateScale(
              marginLeft || defaultMargin,
              defaultScale,
            ),
          }}
        />
      )}
      {sourceUri && (
        <AImage
          sourceUri={sourceUri}
          width={size}
          height={size}
          backgroundColor={backgroundColor}
          resizeMode={resizeMode}
          marginTop={marginTop}
          marginRight={marginRight}
          marginBottom={marginBottom}
          marginLeft={marginLeft}
          cache={cache}
          borderRadius={radiusValue}
        />
      )}
      {source && (
        <AImage
          source={source}
          width={size}
          height={size}
          backgroundColor={backgroundColor}
          resizeMode={resizeMode}
          marginTop={marginTop}
          marginRight={marginRight}
          marginBottom={marginBottom}
          marginLeft={marginLeft}
          borderRadius={radiusValue}
        />
      )}
    </>
  );
};

const styles = (props: {colors: any}) =>
  StyleSheet.create({
    labelStyle: {
      overflow: 'hidden',
    },
  });

export default withTheme(AAvatar);
