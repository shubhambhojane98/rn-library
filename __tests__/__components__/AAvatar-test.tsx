import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import AAvatar from '../../src/components/AAvatar/AAvatar';

it('renders correctly with label', () => {
  const tree = renderer.create(<AAvatar label="AA" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with label, fontSize & radius', () => {
  const tree = renderer
    .create(<AAvatar label="AA" fontSize={24} size={72} borderRadius={36} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with sourceUri', () => {
  const tree = renderer
    .create(<AAvatar sourceUri="https://unsplash.it/400/400?image=1" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
