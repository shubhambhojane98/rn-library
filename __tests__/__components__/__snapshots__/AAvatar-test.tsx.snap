// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with label 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Mulish-Regular",
        "fontSize": 21.13142857142857,
        "fontWeight": "normal",
      },
      Object {
        "color": "#231F20",
      },
      Object {
        "textAlign": "center",
      },
      Object {
        "backgroundColor": "#D7D7D7",
        "borderRadius": 32,
        "height": 64,
        "lineHeight": 64,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "overflow": "hidden",
        "width": 64,
      },
    ]
  }
>
  AA
</Text>
`;

exports[`renders correctly with label, fontSize & radius 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Mulish-Regular",
        "fontSize": 29.485714285714288,
        "fontWeight": "normal",
      },
      Object {
        "color": "#231F20",
      },
      Object {
        "textAlign": "center",
      },
      Object {
        "backgroundColor": "#D7D7D7",
        "borderRadius": 36,
        "height": 72,
        "lineHeight": 72,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "overflow": "hidden",
        "width": 72,
      },
    ]
  }
>
  AA
</Text>
`;

exports[`renders correctly with sourceUri 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 32,
        "height": 78.62857142857143,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "width": 78.62857142857143,
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "cache": undefined,
        "priority": "normal",
        "uri": "https://unsplash.it/400/400?image=1",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`renders correctly with sourceUri, bgColor, size, radius & cache 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "red",
        "borderRadius": 100,
        "height": 245.71428571428572,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "width": 245.71428571428572,
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "cache": "web",
        "priority": "normal",
        "uri": "https://unsplash.it/400/400?image=9",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
